### DISCUSSION TITLE
Paymaster: validate transaction before pay fees [ETHBarcelona]

### DISCUSSION THREAD

**Discussion by seddik11** [zkSync-Community-Hub/zksync-developers#54] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

We are building a Dao and we want to subsidze fees for voting proposals so we incentivize users to vote. We are implementing this solution using Paymaster

### What is the specific issue or error you're encountering?

We want to validate the transaction data we are sponsoring (like transaction.to, function signature etc..) but I don't know the way to do it. Should we passed to paymasterInput ? and which type ("General" or "ApprovalBased") ? also those arguments are passed off-chain, is the Paymaster going to verify this ?

### Can you share the error messages or logs you're receiving, if any?

Not implemented yet

### Have you made any recent changes to the contract before encountering this issue?

No

### Are there any external libraries or contracts that your contract interacts with?

No

### Can you provide the relevant portions of your contract code where the issue is occurring?

```
(address token, uint256 amount, bytes memory data) = abi.decode(
                _transaction.paymasterInput[4:],
                (address, uint256, bytes)
            );
```

Want to retrive data like this and check it

### Have you tried to isolate the problem, and if so, what were the results?

No

### What steps have you already taken to try to resolve the issue?

none

**Comment by dutterbutter** [comment, marked as answer]
Hello 👋 

Not exactly sure the requirements for what your specific validation flow would look like, however, I hope the below explanation helps. For determining what paymaster flow to use:

(1) Use the General based flow if no prior actions are required from the user for the paymaster to operate.
(2) Use the Approval based flow if the user is required to set certain allowance to a token for the paymaster to operate. For example, if you would like the option for users to be able to pay the fees using an ERC20 token. 

Please refer to the docs 👉 [here](https://era.zksync.io/docs/reference/concepts/account-abstraction.html#built-in-paymaster-flows) for more detailed info:

In terms of validation, you will need to handle this logic in the `validateAndPayForPaymasterTransaction` function. Here you have access to `Transaction` struct which provides you the below fields. You can then create conditions based on the `to` field or if the user address owns a particular token or other conditions. You can see some of these flows in our paymaster-examples repo 👉  [here](https://github.com/matter-labs/paymaster-examples/). 

**Example validation check:**
```
// We verify that the user is in the allowList
address userAddress = address(uint160(_transaction.from));

// checks if account is in allowList
bool isAllowed = allowList[userAddress];
require(isAllowed, "Account is not in allow list");
```

**Available tx fields:**
```
struct Transaction {
    // The type of the transaction.
    uint256 txType;
    // The caller.
    uint256 from;
    // The callee.
    uint256 to;
    // The gasLimit to pass with the transaction.
    // It has the same meaning as Ethereum's gasLimit.
    uint256 gasLimit;
    // The maximum amount of gas the user is willing to pay for a byte of pubdata.
    uint256 gasPerPubdataByteLimit;
    // The maximum fee per gas that the user is willing to pay.
    // It is akin to EIP1559's maxFeePerGas.
    uint256 maxFeePerGas;
    // The maximum priority fee per gas that the user is willing to pay.
    // It is akin to EIP1559's maxPriorityFeePerGas.
    uint256 maxPriorityFeePerGas;
    // The transaction's paymaster. If there is no paymaster, it is equal to 0.
    uint256 paymaster;
    // The nonce of the transaction.
    uint256 nonce;
    // The value to pass with the transaction.
    uint256 value;
    // In the future, we might want to add some
    // new fields to the struct. The `txData` struct
    // is to be passed to account and any changes to its structure
    // would mean a breaking change to these accounts. In order to prevent this,
    // we should keep some fields as "reserved".
    // It is also recommended that their length is fixed, since
    // it would allow easier proof integration (in case we will need
    // some special circuit for preprocessing transactions).
    uint256[4] reserved;
    // The transaction's calldata.
    bytes data;
    // The signature of the transaction.
    bytes signature;
    // The properly formatted hashes of bytecodes that must be published on L1
    // with the inclusion of this transaction. Note, that a bytecode has been published
    // before, the user won't pay fees for its republishing.
    bytes32[] factoryDeps;
    // The input to the paymaster.
    bytes paymasterInput;
    // Reserved dynamic type for the future use-case. Using it should be avoided,
    // But it is still here, just in case we want to enable some additional functionality.
    bytes reservedDynamic;
}
```

**Reply by seddik11** [reply]
Thank you. It helps a lot

**Comment by elhajin** [comment]


**Reply by dutterbutter** [reply]
Can you open a separate discussion for the question please as this thread is already marked.

**Comment by caydiverlarachel** [comment]
To validate transaction data sponsored by a Paymaster, you can pass the necessary information (such as transaction to address and function signature) in the paymasterInput field. The type of Paymaster you choose ("General" or "ApprovalBased") depends on your specific requirements and use case. Off-chain verification of the transaction data by the Paymaster is typically not performed; it's usually handled on-chain within your contract logic. You can retrieve and verify the data passed to the Paymaster using the paymasterInput field in your contract.

**Comment by pobertester** [comment]
Environment: Testnet
Issue: Need to

**Comment by cahillnick** [comment]
To validate transaction data when using a Paymaster for subsidizing fees in your DAO, you typically need to pass the relevant transaction details to the Paymaster contract. This can include information such as the transaction destination, function signature, and any additional parameters required for validation.

You can pass this information to the Paymaster either directly as part of the transaction data or as part of the Paymaster input parameters, depending on your implementation preferences. The Paymaster contract then needs to verify this information to ensure that the sponsored transactions meet the desired criteria.

If you're unsure about the specific implementation details or which type of Paymaster (General or ApprovalBased) to use, you may need to refer to the documentation of the Paymaster contract you're using or consult with the developers who implemented it.

Overall, the key is to ensure that the Paymaster contract has the necessary functionality to validate transaction data and enforce any requirements for subsidized transactions in your DAO.

**Comment by dariuscandfield** [comment]
To validate transaction data in your Paymaster contract, you can pass relevant information such as transaction destination address, function signature, and other parameters to the PaymasterInput field. Depending on your validation requirements, you can choose between "General" or "ApprovalBased" types. Ensure that your Paymaster contract includes logic to verify the transaction data passed via PaymasterInput and enforce any necessary validation rules. Testing and verifying this logic thoroughly is essential before deployment.

**Comment by marcellusjustin** [comment]
The user is working on a DAO and wants to subsidize fees for voting proposals using a Paymaster. They want to validate the transaction data they are sponsoring, such as transaction destination and function signature. They are unsure if they should pass this information to the PaymasterInput and which type of input to use (General or ApprovalBased). They are also unsure if the Paymaster will verify this information. They have not made any recent changes to the contract, and there are no external libraries or contracts being used. They are trying to retrieve the data and check it, but have not tried to isolate the problem or take any steps to resolve it yet.

**Comment by cioprogephen** [comment]
The user is building a Dao and wants to subsidize fees for voting proposals using Paymaster on the zkSync Testnet. They want to validate the transaction data they are sponsoring, such as transaction.to and function signature, but are unsure of the way to do it. They are also unsure if they should pass it to paymasterInput and which type (General or ApprovalBased) to use. The arguments are passed off-chain, and the user wants to know if the Paymaster will verify them. The user has provided relevant portions of their contract code where the issue is occurring. The user has not tried to isolate the problem and has not taken any steps to try to resolve the issue. No repo link is provided.
